### PR TITLE
Make task scheduler TrySubmit always non-blocking

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Main idea is to make scheduler.TrySubmit() always non-blocking by allow TrySubmit on the underlying worker pool as well.
- But then the interface of task Processor and task Scheduler becomes exactly the same.
- So removed the Processor interface. The existing ParallelProcessor can be regarded as a FIFO scheduler. 

**Why?**
- Make task scheduler TrySubmit always non-blocking
- Currently if there's only one outstanding task in iwrr scheduler, it will try to do a blocking submit to the worker pool, this leads to a deadlock issue during testing: all task processing goroutines are blocked on a mutex when adding a task to the rescheduler, but the rescheduler hold the mutex, calls TrySubmit to reschedule a task. But that TrySubmit call may be the first outstanding task and blocked on Submitting task to the worker pool. 
- ^ sounds like an extremely rare case, but happens on test cluster, twice...

**How did you test it?**
- Tested on test cluster

**Potential risks**
- Mostly refactoring.

**Is hotfix candidate?**
- no. the case is very rare

### CODE REVIEW

File: common/tasks/fifo_scheduler.go [hunk 1]
```diff
@@ -34,173 +34,183 @@ import (
...
 
 	go func() {
-		if success := common.AwaitWaitGroup(&p.shutdownWG, time.Minute); !success {
-			p.logger.Warn("parallel processor timed out waiting for workers")
+		if success := common.AwaitWaitGroup(&f.shutdownWG, time.Minute); !success {
+			f.logger.Warn("parallel processor timed out waiting for workers")
 		}
 	}()
-	p.logger.Info("parallel processor stopped")
+	f.logger.Info("parallel processor stopped")
```
Comment (yiminc) [inline common/tasks/fifo_scheduler.go:117]: nit: update fifo scheduler

File: common/tasks/interleaved_weighted_round_robin.go [hunk 2]
```diff
@@ -154,8 +150,7 @@ func (s *InterleavedWeightedRoundRobinScheduler[T, K]) TrySubmit(
 	task T,
 ) bool {
 	numTasks := atomic.AddInt64(&s.numInflightTask, 1)
-	if numTasks == 1 {
-		s.doDispatchTasksDirectly(task)
+	if numTasks == 1 && s.tryDispatchTaskDirectly(task) {
```
Comment (yiminc) [inline common/tasks/interleaved_weighted_round_robin.go:153]: i thought this PR is to remove this optimization?

File: service/history/queues/fifo_scheduler.go [hunk 3]
```diff
@@ -47,52 +41,45 @@ type (
 	// and always schedule tasks in fifo order regardless
 	// which namespace the task belongs to.
 	FIFOScheduler struct {
-		wRRScheduler tasks.Scheduler[Executable]
+		scheduler tasks.Scheduler[Executable]
```
Comment (yiminc) [inline service/history/queues/fifo_scheduler.go:44]: what is the difference of this FIFOScheduler and the one under common/tasks/fifo.scheduler.go?

File: common/tasks/interleaved_weighted_round_robin.go (hunk 2 above)
Comment (yycptt) [reply common/tasks/interleaved_weighted_round_robin.go:153]: Removing this optimization is my quick hack for testing multi cursor implementation.
This optimization does make sense to me so I don't want to simply remove it. 

The effort to keep it and fix the issue is not huge and most of this PR is actually doing refactoring for consolidating `Scheduler` and `Processor` interface after adding `TrySubmit` to `Processor` interface.

File: service/history/queues/fifo_scheduler.go (hunk 3 above)
Comment (yycptt) [reply service/history/queues/fifo_scheduler.go:44]: There's no difference actually. But the Scheduler interface defined in `history/queues` package is different from the one in `common/tasks`, so need an adapter. 

Re. the actual difference of the interface, please check this [PR](https://github.com/temporalio/temporal/pull/3266/files#diff-36d52beeb8ce1d18e54aa8aefb9352461e6245bc0d693b397d80ea829f4ac4cfR61). The definition here needs to be updated, and I will update this PR after the linked one is landed.

Comment (yycptt) [reply service/history/queues/fifo_scheduler.go]: Updated. This file is now gone. 

The scheduler defined in `history/queues` package is a thin wrapper around the one defined in `common/tasks`, to support getting the func for task chan and weight assignment.

**Review by yiminc — APPROVED** [review]

**Pull request merged**